### ISSUE TITLE
Compiling to wasm: error: 'string.h' file not found

### ISSUE THREAD

**Issue by cc-marco** [Gekkio/imgui-rs#342]
I am using imgui in wgpu-rs and works very nicely! I am currently writing my engine in a way that works both native and wasm.
As of now I can compile and run just fine on native, as soon as I try to compile to wasm imgui-sys seems to having issues finding string.h. 
I am fairly new to rust and rust ecosystem so I hope that is not a silly problem. 
Here the full error:
```
   Compiling imgui-sys v0.4.0
error: failed to run custom build command for `imgui-sys v0.4.0`

Caused by:
  process didn't exit successfully: `C:\Users\marco\Downloads\imgui-wgpu-rs-master\imgui-wgpu-rs-master\target\debug\build\imgui-sys-19551f715ef222d8\build-script-build` (exit code: 1)
--- stdout
TARGET = Some("wasm32-unknown-unknown")
OPT_LEVEL = Some("0")
HOST = Some("x86_64-pc-windows-msvc")
CXX_wasm32-unknown-unknown = None
CXX_wasm32_unknown_unknown = None
TARGET_CXX = None
CXX = None
CXXFLAGS_wasm32-unknown-unknown = None
CXXFLAGS_wasm32_unknown_unknown = None
TARGET_CXXFLAGS = None
CXXFLAGS = None
CRATE_CC_NO_DEFAULTS = None
DEBUG = Some("true")
CXX_wasm32-unknown-unknown = None
CXX_wasm32_unknown_unknown = None
TARGET_CXX = None
CXX = None
CXXFLAGS_wasm32-unknown-unknown = None
CXXFLAGS_wasm32_unknown_unknown = None
TARGET_CXXFLAGS = None
CXXFLAGS = None
CRATE_CC_NO_DEFAULTS = None
running: "clang" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "--target=wasm32-unknown-unknown" "-Wall" "-Wextra" "-DCIMGUI_NO_EXPORT" "-DIMGUI_DISABLE_WIN32_FUNCTIONS" "-DIMGUI_DISABLE_OSX_FUNCTIONS" "-o" "C:\\Users\\marco\\Downloads\\imgui-wgpu-rs-master\\imgui-wgpu-rs-master\\target\\wasm32-unknown-unknown\\debug\\build\\imgui-sys-00378023ca555e70\\out\\third-party/cimgui/cimgui.o" "-c" "third-party/cimgui/cimgui.cpp"
cargo:warning=In file included from third-party/cimgui/cimgui.cpp:4:
cargo:warning=third-party/cimgui/./imgui/imgui.h:58:10: fatal error: 'string.h' file not found
cargo:warning=#include <string.h>                 // memset, memmove, memcpy, strlen, strchr, strcpy, strcmp
cargo:warning=         ^~~~~~~~~~
cargo:warning=1 error generated.
exit code: 1

--- stderr


error occurred: Command "clang" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "--target=wasm32-unknown-unknown" "-Wall" "-Wextra" "-DCIMGUI_NO_EXPORT" "-DIMGUI_DISABLE_WIN32_FUNCTIONS" "-DIMGUI_DISABLE_OSX_FUNCTIONS" "-o" "C:\\Users\\marco\\Downloads\\imgui-wgpu-rs-master\\imgui-wgpu-rs-master\\target\\wasm32-unknown-unknown\\debug\\build\\imgui-sys-00378023ca555e70\\out\\third-party/cimgui/cimgui.o" "-c" "third-party/cimgui/cimgui.cpp" with args "clang" did not execute successfully (status code exit code: 1).



warning: build failed, waiting for other jobs to finish...
error: build failed
```

I have replicated this both on windows and on macOS, with the same exact error. 

I am simply compiling with :
```
cargo build --target v
```

Does anyone know what the issue might be?
Would be amazing to be able to target wasm as it is possible to do from c++. It would open so many possibilities. 

M.

**Comment by cc-marco** [comment]
As suspected it was a silly mistake on my side, by digging around I found that the crate exposes a wasm feature. By enabling that feature in my cargo I was able to successfully.
Although I am still unable to run, the wasm-bindgen seems to generate a js which tries to import "imgui-sys-v0" which I am not sure where it is coming from. 

```js
import * as __wbg_star0 from 'imgui-sys-v0';

let wasm;

const heap = new Array(32).fill(undefined);

heap.push(undefined, null, true, false);

function getObject(idx) { return heap[idx]; }
...
```

where later on ```__wbg_star0``` seems to be added to the imports for the wasm module. 

Anyone has any idea?

**Comment by sumeet** [comment]
hi @cc-marco, were you able to get it working? i'm interested in using imgui-rs inside of a browser, and not sure where to get started

**Comment by 4bb4** [comment]
@cc-marco I've spent a good amount of time looking into this, and it seems like that import comes from the `#[link(wasm_import_module = "imgui-sys-v0")]` decorators in `imgui-sys/src/wasm_bindings.rs`. I've looked at this more, and this seems to be a mechanism to link to external JS modules, at least this is what would be required for an import statement like the one in your code example (I get the same when running `wasm-pack build --target web`). 

There are JS bindings for imgui (https://github.com/flyover/imgui-js), but they are set up in a way as to be run with Node, and I'm not very experienced with these things so I could not get anything to run yet. Another interesting observation is that that WASM binding code seems to have come from someone at Embark Studios, judging from the git blame. We could try pinging the author of the related PRs (https://github.com/Gekkio/imgui-rs/pull/323/commits/d638841e69dda9244d39e23782f192f42c78c0af) - they are very likely to know what's going on here, though I'm afraid this may relate to internal tools they have.

**Comment by 1danielcoelho** [comment]
I've compiled a local copy of `imgui-rs` undoing that PR (so just naively renaming "imgui-sys-v0" to "imgui-sys"), and that doesn't fix it though: The error message just turns into this instead:
```
ERROR in ./pkg/index_bg.wasm
Module not found: Error: Can't resolve 'imgui-sys' in '<path>'
 @ ./pkg/index_bg.wasm
 @ ./pkg/index.js
 @ ./index.js
```

**Comment by 4bb4** [comment]
@1danielcoelho I'm quite sure that won't be enough, it really seems like one has to create or at least provide a WASM / JS implementation of imgui itself (created with emscripten, for example). I have not seen any building code in the imgui-rs repo for the WASM path, in fact in `https://github.com/Gekkio/imgui-rs/blob/master/imgui-sys/build.rs`, you can see that specifically the `cc` crate is only used to compile the C++ code of imgui itself if WASM is _not_ the target. 

@Gekkio do you know more about this, or were you not really involved in the development of that WASM compatibility part?

**Comment by expenses** [comment]
Hey @Jasper-Bekkers, where do we get this 'imgui-sys-v0' import js file from?

**Comment by dbr** [comment]
> Did anyone manage the imgui-rs + WASM combo to work?

Yes, there is some information in #447 about it, but it seems far from simple

However this was about a year ago, and the WASM stuff is moving quite quickly, so the state is likely quite different now
